File: src/ObjectManager/Object.Tes/FilePacks/Records/045-REFR.Placed Object.cs
using OA.Core;
using System.Collections.Generic;
using UnityEngine;

namespace OA.Tes.FilePacks.Records
{
    public class REFRRecord : Record
    {
        public struct XTELField
        {
            public FormId<REFRRecord> Door;
            public Vector3 Position;
            public Vector3 Rotation;

            public XTELField(UnityBinaryReader r, int dataSize)
            {
                Door = new FormId<REFRRecord>(r.ReadLEUInt32());
                Position = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
                Rotation = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
            }
        }

        public struct DATAField
        {
            public Vector3 Position;
            public Vector3 Rotation;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                Position = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
                Rotation = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
            }
        }

        public struct XLOCField
        {
            public override string ToString() => $"{Key}";
            public byte LockLevel;
            public FormId<KEYMRecord> Key;
            public byte Flags;

            public XLOCField(UnityBinaryReader r, int dataSize)
            {
                LockLevel = r.ReadByte();
                r.SkipBytes(3); // Unused
                Key = new FormId<KEYMRecord>(r.ReadLEUInt32());
                if (dataSize == 16)
                    r.SkipBytes(4); // Unused
                Flags = r.ReadByte();
                r.SkipBytes(3); // Unused
            }
        }

        public struct XESPField
        {
            public override string ToString() => $"{Reference}";
            public FormId<Record> Reference;
            public byte Flags;

            public XESPField(UnityBinaryReader r, int dataSize)
            {
                Reference = new FormId<Record>(r.ReadLEUInt32());
                Flags = r.ReadByte();
                r.SkipBytes(3); // Unused
            }
        }

        public struct XSEDField
        {
            public override string ToString() => $"{Seed}";
            public byte Seed;

            public XSEDField(UnityBinaryReader r, int dataSize)
            {
                Seed = r.ReadByte();
                if (dataSize == 4)
                    r.SkipBytes(3); // Unused
            }
        }

        public class XMRKGroup
        {
            public override string ToString() => $"{FULL.Value}";
            public BYTEField FNAM; // Map Flags
            public STRVField FULL; // Name
            public BYTEField TNAM; // Type
        }

        public override string ToString() => $"REFR: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public FMIDField<Record> NAME; // Base
        public XTELField? XTEL; // Teleport Destination (optional)
        public DATAField DATA; // Position/Rotation
        public XLOCField? XLOC; // Lock information (optional)
        public List<CELLRecord.XOWNGroup> XOWNs; // Ownership (optional)
        public XESPField? XESP; // Enable Parent (optional)
        public FMIDField<Record>? XTRG; // Target (optional)
        public XSEDField? XSED; // SpeedTree (optional)
        public BYTVField? XLOD; // Distant LOD Data (optional)
        public FLTVField? XCHG; // Charge (optional)
        public FLTVField? XHLT; // Health (optional)
        public FMIDField<CELLRecord>? XPCI; // Unused (optional)
        public IN32Field? XLCM; // Level Modifier (optional)
        public FMIDField<REFRRecord>? XRTM; // Unknown (optional)
        public UI32Field? XACT; // Action Flag (optional)
        public IN32Field? XCNT; // Count (optional)
        public List<XMRKGroup> XMRKs; // Ownership (optional)
        //public bool? ONAM; // Open by Default
        public BYTVField? XRGD; // Ragdoll Data (optional)
        public FLTVField? XSCL; // Scale (optional)
        public BYTEField? XSOL; // Contained Soul (optional)
        int _nextFull;

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "NAME": NAME = new FMIDField<Record>(r, dataSize); return true;
                case "XTEL": XTEL = new XTELField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                case "XLOC": XLOC = new XLOCField(r, dataSize); return true;
                case "XOWN": if (XOWNs == null) XOWNs = new List<CELLRecord.XOWNGroup>(); XOWNs.Add(new CELLRecord.XOWNGroup { XOWN = new FMIDField<Record>(r, dataSize) }); return true;
                case "XRNK": ArrayUtils.Last(XOWNs).XRNK = new IN32Field(r, dataSize); return true;
                case "XGLB": ArrayUtils.Last(XOWNs).XGLB = new FMIDField<Record>(r, dataSize); return true;
                case "XESP": XESP = new XESPField(r, dataSize); return true;
                case "XTRG": XTRG = new FMIDField<Record>(r, dataSize); return true;
                case "XSED": XSED = new XSEDField(r, dataSize); return true;
                case "XLOD": XLOD = new BYTVField(r, dataSize); return true;
                case "XCHG": XCHG = new FLTVField(r, dataSize); return true;
                case "XHLT": XCHG = new FLTVField(r, dataSize); return true;
                case "XPCI": XPCI = new FMIDField<CELLRecord>(r, dataSize); _nextFull = 1; return true;
                case "FULL":
                    if (_nextFull == 1) XPCI.Value.AddName(r.ReadASCIIString(dataSize));
                    else if (_nextFull == 2) ArrayUtils.Last(XMRKs).FULL = new STRVField(r, dataSize);
                    _nextFull = 0;
                    return true;
                case "XLCM": XLCM = new IN32Field(r, dataSize); return true;
                case "XRTM": XRTM = new FMIDField<REFRRecord>(r, dataSize); return true;
                case "XACT": XACT = new UI32Field(r, dataSize); return true;
                case "XCNT": XCNT = new IN32Field(r, dataSize); return true;
                case "XMRK": if (XMRKs == null) XMRKs = new List<XMRKGroup>(); XMRKs.Add(new XMRKGroup()); _nextFull = 2; return true;
                case "FNAM": ArrayUtils.Last(XMRKs).FNAM = new BYTEField(r, dataSize); return true;
                case "TNAM": ArrayUtils.Last(XMRKs).TNAM = new BYTEField(r, dataSize); r.ReadByte(); return true;
                case "ONAM": return true;
                case "XRGD": XRGD = new BYTVField(r, dataSize); return true;
                case "XSCL": XSCL = new FLTVField(r, dataSize); return true;
                case "XSOL": XSOL = new BYTEField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-REFR.Placed Object.swift
//
//  REFRRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//
import simd

public class REFRRecord: Record {
    public struct XTELField {
        public let door: FormId<REFRRecord>
        public let position: float3
        public let rotation: float3

        init(_ r: BinaryReader, _ dataSize: Int) {
            door = FormId<REFRRecord>(r.readLEUInt32())
            position = r.readLEFloat3()
            rotation = r.readLEFloat3()
        }
    }

    public struct DATAField {
        public let position: float3
        public let rotation: float3

        init(_ r: BinaryReader, _ dataSize: Int) {
            position = r.readLEFloat3()
            rotation = r.readLEFloat3()
        }
    }

    public struct XLOCField: CustomStringConvertible {
        public var description: String { return "\(key)" }
        public let lockLevel: UInt8
        public let key: FormId<KEYMRecord>
        public let flags: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            lockLevel = r.readByte()
            r.skipBytes(3) // Unused
            key = FormId<KEYMRecord>(r.readLEUInt32())
            if dataSize == 16 {
                r.skipBytes(4) // Unused
            }
            flags = r.readByte()
            r.skipBytes(3) // Unused
        }
    }

    public struct XESPField: CustomStringConvertible {
        public var description: String { return "\(reference)" }
        public let reference: FormId<Record>
        public let flags: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            reference = FormId<Record>(r.readLEUInt32())
            flags = r.readByte()
            r.skipBytes(3) // Unused
        }
    }

    public struct XSEDField: CustomStringConvertible {
        public var description: String { return "\(seed)" }
        public let seed: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            seed = r.readByte()
            if dataSize == 4 {
                r.skipBytes(3) // Unused
            }
        }
    }

    public class XMRKGroup: CustomStringConvertible {
        public var description: String { return "\(FULL!)" }
        public var FNAM: BYTEField! // Map Flags
        public var FULL: STRVField! // Name
        public var TNAM: BYTEField! // Type
    }

    public override var description: String { return "REFR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var NAME: FMIDField<Record>! // Base
    public var XTEL: XTELField? = nil// Teleport Destination (optional)
    public var DATA: DATAField! // Position/Rotation
    public var XLOC: XLOCField? = nil // Lock information (optional)
    public var XOWNs: [CELLRecord.XOWNGroup]? = nil // Ownership (optional)
    public var XESP: XESPField? = nil // Enable Parent (optional)
    public var XTRG: FMIDField<Record>? = nil // Target (optional)
    public var XSED: XSEDField? = nil // SpeedTree (optional)
    public var XLOD: BYTVField? = nil // Distant LOD Data (optional)
    public var XCHG: FLTVField? = nil // Charge (optional)
    public var XHLT: FLTVField? = nil // Health (optional)
    public var XPCI: FMIDField<CELLRecord>? = nil // Unused (optional)
    public var XLCM: IN32Field? = nil // Level Modifier (optional)
    public var XRTM: FMIDField<REFRRecord>? = nil // Unknown (optional)
    public var XACT: UI32Field? = nil // Action Flag (optional)
    public var XCNT: IN32Field? = nil // Count (optional)
    public var XMRKs: [XMRKGroup]? = nil // Ownership (optional)
    //public var ONAM: Bool? // Open by Default
    public var XRGD: BYTVField? = nil // Ragdoll Data (optional)
    public var XSCL: FLTVField? = nil // Scale (optional)
    public var XSOL: BYTEField? = nil // Contained Soul (optional)
    var _nextFull: Int!
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "NAME": NAME = FMIDField<Record>(r, dataSize)
        case "XTEL": XTEL = XTELField(r, dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        case "XLOC": XLOC = XLOCField(r, dataSize)
        case "XOWN": if XOWNs == nil { XOWNs = [CELLRecord.XOWNGroup]() }; XOWNs!.append(CELLRecord.XOWNGroup(XOWN: FMIDField<Record>(r, dataSize)))
        case "XRNK": XOWNs!.last!.XRNK = r.readT(dataSize)
        case "XGLB": XOWNs!.last!.XGLB = FMIDField<Record>(r, dataSize)
        case "XESP": XESP = XESPField(r, dataSize)
        case "XTRG": XTRG = FMIDField<Record>(r, dataSize)
        case "XSED": XSED = XSEDField(r, dataSize)
        case "XLOD": XLOD = r.readBYTV(dataSize)
        case "XCHG": XCHG = r.readT(dataSize)
        case "XHLT": XCHG = r.readT(dataSize)
        case "XPCI": XPCI = FMIDField<CELLRecord>(r, dataSize); _nextFull = 1
        case "FULL":
            if _nextFull == 1 { XPCI!.add(name: r.readASCIIString(dataSize)) }
            else if _nextFull == 2 { XMRKs!.last!.FULL = r.readSTRV(dataSize) }
            _nextFull = 0
        case "XLCM": XLCM = r.readT(dataSize)
        case "XRTM": XRTM = FMIDField<REFRRecord>(r, dataSize)
        case "XACT": XACT = r.readT(dataSize)
        case "XCNT": XCNT = r.readT(dataSize)
        case "XMRK": if XMRKs == nil { XMRKs = [XMRKGroup]() }; XMRKs!.append(XMRKGroup()); _nextFull = 2
        case "FNAM": XMRKs!.last!.FNAM = r.readT(dataSize)
        case "TNAM": XMRKs!.last!.TNAM = r.readT(dataSize) //; r.skipBytes(1)
        case "ONAM": break
        case "XRGD": XRGD = r.readBYTV(dataSize)
        case "XSCL": XSCL = r.readT(dataSize)
        case "XSOL": XSOL = r.readT(dataSize)
        default: return false
        }
        return true
    }
}
